### PR TITLE
Fix /help 2 for paging, Fix tab completion, add Tips.

### PR DESCRIPTION
**Pull request by ryantheleach**
Requires: https://github.com/SpongePowered/SpongeAPI/pull/1539

Alternative to: https://github.com/SpongePowered/SpongeCommon/pull/1303

Fixes: https://github.com/SpongePowered/SpongeCommon/issues/1242

/page is still preferable to /help as /help reconstructs the entire pagination each time. 

/help 2 however will still continue to work even after the pagination expires.

I've also added the default minecraft help tip to the text, and changed the header to more closely match the default minecraft header.

image

image

image

Only current downside, is that the error handling slightly pollutes the usage text for /help.

image

Also now with tab completion.

image

### CODE REVIEW

File: src/main/java/org/spongepowered/common/command/SpongeHelpCommand.java [hunk 1]
```diff
@@ -25,69 +25,130 @@
...
+import java.util.stream.Collectors;
+
+import javax.annotation.Nullable;
 
 public class SpongeHelpCommand {
 
-    private static final Comparator<CommandMapping> COMMAND_COMPARATOR = (o1, o2) -> o1.getPrimaryAlias().compareTo(o2.getPrimaryAlias());
+    private static final Comparator<CommandMapping> COMMAND_COMPARATOR = Comparator.comparing(CommandMapping::getPrimaryAlias);
+
+    private static final Text pageKey = Text.of("page");
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/command/SpongeHelpCommand.java]: `PAGE_KEY`

File: src/main/java/org/spongepowered/common/command/SpongeHelpCommand.java [hunk 2]
```diff
@@ -25,69 +25,130 @@
...
+    private static final Text commandKey = Text.of("command");
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/command/SpongeHelpCommand.java]: `COMMAND_KEY`

File: src/main/java/org/spongepowered/common/command/SpongeHelpCommand.java [hunk 3]
```diff
@@ -25,69 +25,130 @@
...
+    private static final Text notFound = Text.of("notFound");
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/command/SpongeHelpCommand.java]: `NOT_FOUND`

File: src/main/java/org/spongepowered/common/command/SpongeHelpCommand.java [hunk 4]
```diff
@@ -25,69 +25,130 @@
...
+
+    private static final CommandElement commandArgument = new CommandElement(commandKey){
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/command/SpongeHelpCommand.java]: `COMMAND_ARGUMENT`

**Review by kashike — CHANGES_REQUESTED** [review]

File: src/main/java/org/spongepowered/common/command/SpongeHelpCommand.java [hunk 5]
```diff
@@ -25,75 +25,134 @@
...
 public class SpongeHelpCommand {
 
-    private static final Comparator<CommandMapping> COMMAND_COMPARATOR = (o1, o2) -> o1.getPrimaryAlias().compareTo(o2.getPrimaryAlias());
+    private static final Comparator<CommandMapping> COMMAND_COMPARATOR = Comparator.comparing(CommandMapping::getPrimaryAlias);
+
+    private static final Text PAGE_KEY = Text.of("page");
+    private static final Text COMMAND_KEY = Text.of("command");
+    private static final Text NOT_FOUND = Text.of("notFound");
+
+    private static final CommandElement COMMAND_ARGUMENT = new CommandElement(COMMAND_KEY){
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/command/SpongeHelpCommand.java]: `) {`

**Comment by Zidane** [comment]
@ryantheleach 

Is this still needed? If so, get this rebased on latest bleeding please.

**Comment by ryantheleach** [comment]
https://github.com/SpongePowered/SpongeCommon/pull/1475 covered adding the methods needed, but this still has changes to the HelpCommand itself.

Seems like a small task to fix it and get ready, I'll get this done tonight.

**Comment by ryantheleach** [comment]
Apologies for the API Bump/Revert API Bump, was having a little trouble with the rebase.

**Comment by Zidane** [comment]
@ryantheleach 

Do NOT separate the Help command back into its own class. That was merged in Grinch's PR for the plugins command changes due to my specific instructions.

File: src/main/java/org/spongepowered/common/command/SpongeCommandGenerator.java [hunk 6]
```diff
@@ -737,22 +770,38 @@ private static Long mean(long[] values) {
...
-            .extendedDescription(
-                    Text.of("View a list of all commands. Hover over\n" + " a command to view its description. Click\n"
-                            + " a command to insert it into your chat bar."))
-            .executor((src, args) -> {
-                Optional<String> command = args.getOne("command");
-                if (command.isPresent()) {
-                    Optional<? extends CommandMapping> mapping = SpongeImpl.getGame().getCommandManager().get(command.get(), src);
-                    if (mapping.isPresent()) {
-                        CommandCallable callable = mapping.get().getCallable();
+                .builder()
```
Comment (ryantheleach) [inline src/main/java/org/spongepowered/common/command/SpongeCommandGenerator.java]: Is this whitespace weird? I notice it differs from grinches.

File: src/main/java/org/spongepowered/common/command/SpongeCommandGenerator.java [hunk 7]
```diff
@@ -109,9 +117,12 @@
 import java.util.Optional;
 import java.util.TreeSet;
 import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import javax.annotation.Nullable;
 
 @NonnullByDefault
-public class SpongeCommands {
+public class SpongeCommandGenerator {
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/command/SpongeCommandGenerator.java]: Why the rename?

Comment (parlough) [reply src/main/java/org/spongepowered/common/command/SpongeCommandGenerator.java]: I think Zidane wanted this.

Comment (kashike) [reply src/main/java/org/spongepowered/common/command/SpongeCommandGenerator.java]: ~~I see no comment about it.~~

Comment (kashike) [reply src/main/java/org/spongepowered/common/command/SpongeCommandGenerator.java]: Use `SpongeCommandFactory` instead.

**Comment by kashike** [comment]
Could you please update this pull request, @ryantheleach?

**Comment by parlough** [comment]
Should be good for a final review @kashike along with its Forge PR: https://github.com/SpongePowered/SpongeForge/pull/1734

Make sure to squash the commits on merge :P, there's kind of a lot of back and forth changes here.

**Review by kashike — APPROVED** [review]

**Pull request merged**